cachito/workers/tasks/utils.py

```diff
@@ -203,3 +205,72 @@ def _get_request_or_fail(request_id: int, connect_error_msg: str, status_error_m
+    """A collection of resolved packages."""
+
+    def __init__(self) -> None:
+        self._index = set()
+        self._packages = []
+
+    def add_package(self, pkg_info: Dict[str, str], path: str, deps: List[Dict[str, Any]]) -> None:
+        """Add a package with deps.
+
+        :param dict[str, str] pkg_info: a mapping containing a package information.
+            It must have ``name``, ``type`` and ``version`` key/value pairs.
+        :param str path: the path where the package is retreived. Consult with the
+            ``fetch_*_source`` for the defailed information about a package's path.
+        :param deps: a list of depencencies the package has.
+        :type deps: list[dict[str, any]]
+        :raises CachitoError: if there is a package with same name, type and version
+            has been added already.
+        """
+        key = (pkg_info["name"], pkg_info["type"], pkg_info["version"])
+        if key in self._index:
+            raise CachitoError(f"Duplicate package is detected: {pkg_info!r}")
+        self._index.add(key)
+        self._packages.append(
+            {
+                "name": pkg_info["name"],
+                "type": pkg_info["type"],
+                "version": pkg_info["version"],
+                "path": path,
+                "dependencies": deps,
+            }
+        )
+
+    def write_to_file(self, file_name: Union[str, Path]) -> None:
+        """Write the added packages to a file as JSON data.
+
+        :param file_name: an absolute or relative filename to write the added packages into.
+            When a relative path is used, it will be opened directly and depends on the
+            ``os.curdir``.
+        :type file_name: str or pathlib.Path
+        """
+        log.debug("Write packages with dependencies into file %s: %r", file_name, self._packages)
+        with open(file_name, "w+", encoding="utf-8") as f:
+            json.dump({"packages": self._packages}, f)
+
+    def load(self, file_name: Union[str, Path]) -> None:
+        """Load data from a specified file written by write_to_file method.
+
+        :param file_name: an absolute or relative filename to write the added packages into.
+            When a relative path is used, it will be opened directly and depends on the
+            ``os.curdir``. If the file does not exist, nothing is changed internally.
+        :type file_name: str or pathlib.Path
+        """
+        if not os.path.exists(file_name):
+            log.debug("No data is loaded from non-existing file %s.", file_name)
+            return
+        with open(file_name, "r", encoding="utf-8") as f:
+            data = json.load(f)
+            for p in data["packages"]:
+                self.add_package(p, p["path"], p["dependencies"])
+
+    def __add__(self, other):
+        new = PackagesData()
+        new._packages = self._packages[:]
+        for p in other._packages:
+            new.add_package(p, p["path"], p["dependencies"])
+        return new
```

Comment:
Could this be changed back to `__iadd__`? And instead of creating a new PackagesData instance every time, just update `self`. The JSON files may get pretty big, copying all the data when merging the files could be expensive